src/librustc/ty/mod.rs

```diff
@@ -1343,49 +1342,98 @@ impl<'tcx> ToPredicate<'tcx> for PolyProjectionPredicate<'tcx> {
     }
 }
 
+// A custom iterator used by Predicate::walk_tys.
+struct WalkTysIterator<'tcx, I, J, K>
+    where I: Iterator<Item = Ty<'tcx>>,
+          J: Iterator<Item = Ty<'tcx>>,
+          K: Iterator<Item = Ty<'tcx>>
+{
+    items: WalkTysItems<'tcx, I, J, K>
+}
+
+enum WalkTysItems<'tcx, I, J, K>
+    where I: Iterator<Item = Ty<'tcx>>,
+          J: Iterator<Item = Ty<'tcx>>,
+          K: Iterator<Item = Ty<'tcx>>
+{
+    None,
+    One(Ty<'tcx>),
+    Two((Ty<'tcx>, Ty<'tcx>)),
+    Types(I),
+    InputTypes(J),
+    ProjectionTypes(K)
+}
+
+impl<'tcx, I, J, K> Iterator for WalkTysIterator<'tcx, I, J, K>
```

Comment:
I think you can just `impl Iterator for WalkTysItems` and get rid of the wrapper struct. Then in your implementation of `fn next()` instead of doing `self.items = ...`, you'd just do `*self = ....`.

Reply:
I wanted to do that, but `self` is borrowed; that approach would require some `mem::replace` gimmicks, which isn't free.

Reply:
That's what I thought but then I tried it [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=81bb21ba4622f3c4288a77ab43c47162) and it seemed to work fine.

If it doesn't work in this situation, that's cool.

Reply:
It totally did work, already updated :).